[JSC] Implement import-assertion and JSON module

Constellation opened the pull request:
#### cebdad8ac8691b85db85bb79f54543cae829ed7c
<pre>
[JSC] Implement import-assertion and JSON module
<a href="https://bugs.webkit.org/show_bug.cgi?id=243622">https://bugs.webkit.org/show_bug.cgi?id=243622</a>

Reviewed by Alexey Shvayka and Ryosuke Niwa.

This patch implements import-assertions and JSON module.

1. Import assertions&apos; syntax work was already done. We propagate assertions through ScriptFetchParameters, which already exists.
2. Get &quot;type&quot; assertion, and based on that, we parse JSON if it is &quot;json&quot;.
3. Add SyntheticModuleRecord (this is described in the spec), and materialize JSON module as SyntheticModuleRecord.
4. Fix MIMETypeRegistry::isSupportedJSONMIMEType. &quot;text/json&quot; is explicitly listed in [1], so we should say true for that.
5. Dynamic import also gets assertions via its second parameter.

One test262 test is failing just because our harness is not handling it well (so this is not a bug in JSON module implementation).
WPT failures are due to old module loader&apos;s behavior (so, fixing it requires update of module loader implementation, not related to
JSON module itself).

[1]: <a href="https://mimesniff.spec.whatwg.org/#json-mime-type">https://mimesniff.spec.whatwg.org/#json-mime-type</a>

* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/css-module/charset-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/css-module/content-type-checking-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/css-module/cors-crossorigin-requests-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/css-module/css-module-worker-test-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/css-module/import-css-module-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/css-module/integrity-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/css-module/load-error-events-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/import-assertions/empty-assertion-clause-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/import-assertions/invalid-type-assertion-error-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/import-assertions/unsupported-assertion-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/charset-bom.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/charset-bom.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/charset-bom.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/charset-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/cors-crossorigin-requests-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/integrity-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/json-module-service-worker-test.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/load-error-events-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/module-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/non-object.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/non-object.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/non-object.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/repeated-imports.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/repeated-imports.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/repeated-imports.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/valid-content-type-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* Source/JavaScriptCore/API/JSAPIGlobalObject.mm:
(JSC::JSAPIGlobalObject::moduleLoaderImportModule):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/ModuleLoader.js:
(requestSatisfy):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::Heap):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/HeapSubspaceTypes.h:
* Source/JavaScriptCore/jsc.cpp:
(GlobalObject::moduleLoaderImportModule):
(GlobalObject::moduleLoaderFetch):
* Source/JavaScriptCore/parser/ModuleAnalyzer.cpp:
(JSC::ModuleAnalyzer::appendRequestedModule):
* Source/JavaScriptCore/parser/ModuleAnalyzer.h:
* Source/JavaScriptCore/parser/NodesAnalyzeModule.cpp:
(JSC::tryCreateAssertion):
(JSC::ImportDeclarationNode::analyzeModule):
(JSC::ExportAllDeclarationNode::analyzeModule):
(JSC::ExportNamedDeclarationNode::analyzeModule):
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser&lt;LexerType&gt;::parseImportAssertions):
* Source/JavaScriptCore/parser/SourceProvider.h:
* Source/JavaScriptCore/runtime/AbstractModuleRecord.cpp:
(JSC::AbstractModuleRecord::appendRequestedModule):
(JSC::AbstractModuleRecord::link):
(JSC::AbstractModuleRecord::evaluate):
(JSC::AbstractModuleRecord::dump):
* Source/JavaScriptCore/runtime/AbstractModuleRecord.h:
(JSC::AbstractModuleRecord::requestedModules const):
* Source/JavaScriptCore/runtime/Completion.cpp:
(JSC::retrieveAssertionsFromDynamicImportOptions):
* Source/JavaScriptCore/runtime/Completion.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::syntheticModuleRecordStructure const):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSModuleLoader.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::JSONParse):
(JSC::JSONParseWithException):
* Source/JavaScriptCore/runtime/JSONObject.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/ScriptFetchParameters.h:
(JSC::ScriptFetchParameters::ScriptFetchParameters):
(JSC::ScriptFetchParameters::type const):
(JSC::ScriptFetchParameters::integrity const):
(JSC::ScriptFetchParameters::isTopLevelModule const):
(JSC::ScriptFetchParameters::create):
(JSC::ScriptFetchParameters::~ScriptFetchParameters): Deleted.
* Source/JavaScriptCore/runtime/SyntheticModuleRecord.cpp: Added.
(JSC::SyntheticModuleRecord::createStructure):
(JSC::SyntheticModuleRecord::create):
(JSC::SyntheticModuleRecord::SyntheticModuleRecord):
(JSC::SyntheticModuleRecord::destroy):
(JSC::SyntheticModuleRecord::finishCreation):
(JSC::SyntheticModuleRecord::visitChildrenImpl):
(JSC::SyntheticModuleRecord::link):
(JSC::SyntheticModuleRecord::evaluate):
(JSC::SyntheticModuleRecord::tryCreateWithExportNamesAndValues):
(JSC::SyntheticModuleRecord::tryCreateDefaultExportSyntheticModule):
(JSC::SyntheticModuleRecord::parseJSONModule):
* Source/JavaScriptCore/runtime/SyntheticModuleRecord.h: Added.
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.cpp:
(JSC::JSWebAssemblyInstance::tryCreate):
* Source/WebCore/bindings/js/CachedModuleScriptLoader.cpp:
(WebCore::CachedModuleScriptLoader::create):
(WebCore::CachedModuleScriptLoader::CachedModuleScriptLoader):
* Source/WebCore/bindings/js/CachedModuleScriptLoader.h:
* Source/WebCore/bindings/js/ModuleScriptLoader.h:
(WebCore::ModuleScriptLoader::parameters):
(WebCore::ModuleScriptLoader::ModuleScriptLoader):
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::loadModuleScriptInWorld):
(WebCore::ScriptController::loadModuleScript):
(WebCore::ScriptController::evaluateModule):
* Source/WebCore/bindings/js/ScriptController.h:
* Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
(WebCore::ScriptModuleLoader::fetch):
(WebCore::ScriptModuleLoader::importModule):
(WebCore::ScriptModuleLoader::notifyFinished):
* Source/WebCore/bindings/js/ScriptModuleLoader.h:
* Source/WebCore/bindings/js/WorkerModuleScriptLoader.cpp:
(WebCore::WorkerModuleScriptLoader::create):
(WebCore::WorkerModuleScriptLoader::WorkerModuleScriptLoader):
* Source/WebCore/bindings/js/WorkerModuleScriptLoader.h:
* Source/WebCore/bindings/js/WorkerScriptFetcher.h:
* Source/WebCore/dom/LoadableModuleScript.cpp:
(WebCore::LoadableModuleScript::LoadableModuleScript):
* Source/WebCore/dom/ModuleFetchParameters.h:
(WebCore::ModuleFetchParameters::create):
(WebCore::ModuleFetchParameters::ModuleFetchParameters):
(WebCore::ModuleFetchParameters::integrity const): Deleted.
(WebCore::ModuleFetchParameters::isTopLevelModule const): Deleted.
* Source/WebCore/platform/MIMETypeRegistry.cpp:
(WebCore::MIMETypeRegistry::isSupportedJSONMIMEType):
* Source/WebCore/workers/WorkerOrWorkletScriptController.cpp:
(WebCore::WorkerOrWorkletScriptController::loadAndEvaluateModule):
* Source/WebCore/workers/WorkerThread.cpp:
(WebCore::WorkerThread::evaluateScriptIfNecessary):
* Tools/TestWebKitAPI/Tests/WebCore/MIMETypeRegistry.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/253234@main">https://commits.webkit.org/253234@main</a>
</pre>

Jarred-Sumner commented on Source/JavaScriptCore/runtime/SyntheticModuleRecord.cpp:
```diff
@@ -0,0 +1,133 @@
+{
+    return jsUndefined();
+}
+
```
To add support for arbitrary export names and values, what about something like this? I haven't tried compiling/running this code
```suggestion
SyntheticModuleRecord* SyntheticModuleRecord::tryCreateWithExportNamesAndValues(JSGlobalObject* globalObject, const Identifier& moduleKey, Vector<Identifier>& exportNames, Vector<JSValue>& exportValues)
{
    VM& vm = globalObject->vm();
    auto scope = DECLARE_THROW_SCOPE(vm);

    auto* moduleRecord = create(globalObject, vm, globalObject->syntheticModuleRecordStructure(), moduleKey);

    SymbolTable* exportSymbolTable = SymbolTable::create(vm);
     {
        auto offset = exportSymbolTable->takeNextScopeOffset(NoLockingNecessary);
        exportSymbolTable->set(NoLockingNecessary, vm.propertyNames->starNamespacePrivateName.impl(), SymbolTableEntry(VarOffset(offset)));
    }

    ASSERT(exportNames.size() == exportValues.size());

    for (size_t i = 0; i < exportNames.size(); ++i) {
        auto exportName = exportNames[i];
        auto exportValue = exportValues[i];
       
        moduleRecord->addExportEntry(ExportEntry::createLocal(exportName, exportName));
        auto offset = exportSymbolTable->takeNextScopeOffset(NoLockingNecessary);
        exportSymbolTable->set(NoLockingNecessary, exportName, SymbolTableEntry(VarOffset(offset)));
    }

    JSModuleEnvironment* moduleEnvironment = JSModuleEnvironment::create(vm, globalObject, nullptr, exportSymbolTable, jsTDZValue(), moduleRecord);
    moduleRecord->setModuleEnvironment(globalObject, moduleEnvironment);
    RETURN_IF_EXCEPTION(scope, { });

    bool shouldThrowReadOnlyError = false;
    bool ignoreReadOnlyErrors = true;
    bool putResult = false;

    for (size_t i = 0; i < exportNames.size(); ++i) {
        JSC::PropertyName exportName(exportNames[i]);
        auto exportValue = exportValues[i];       
        putResult = symbolTablePutTouchWatchpointSet(moduleEnvironment, globalObject, exportName, exportValue, shouldThrowReadOnlyError, ignoreReadOnlyErrors, putResult);
        RETURN_IF_EXCEPTION(scope, { });
        RELEASE_ASSERT(putResult);
    }
    RETURN_IF_EXCEPTION(scope, { });

    return moduleRecord;
}

SyntheticModuleRecord* SyntheticModuleRecord::createWithExportNamesAndValues(JSGlobalObject* globalObject, const Identifier& moduleKey, Vector<Identifier>& exportNames, Vector<JSValue>& exportValues)
{
    VM& vm = globalObject->vm();
    auto scope = DECLARE_THROW_SCOPE(vm);
    RELEASE_AND_RETURN(scope, SyntheticModuleRecord::tryCreateDefaultExportSyntheticModule(globalObject, moduleKey, result));    
}
```

Constellation replied:
That sounds good.

rniwa commented on Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
```diff
@@ -195,12 +195,12 @@ JSC::JSInternalPromise* ScriptModuleLoader::fetch(JSC::JSGlobalObject* jsGlobalO
+    RefPtr<JSC::ScriptFetchParameters> assertions;
     if (auto* scriptFetchParameters = JSC::jsDynamicCast<JSC::JSScriptFetchParameters*>(parameters))
-        topLevelFetchParameters = static_cast<ModuleFetchParameters*>(&scriptFetchParameters->parameters());
+        assertions = &scriptFetchParameters->parameters();
```
Are these assertions or parameters? We should probably stick to one terminology.

Constellation replied:
We should name it as `parameters`, since we can have non-assertion data too. (And dynamic `import` is designed actually to take non assertions in the future).

rniwa commented on Source/JavaScriptCore/API/JSAPIGlobalObject.mm:
```diff
@@ -152,8 +153,26 @@
+        if (!assertions.isEmpty()) {
+            ScriptFetchParameters::Type type = ScriptFetchParameters::Type::None;
+            auto typeString = assertions.get(vm.propertyNames->type.impl());
+            if (typeString == "json"_s)
```
It looks a bit inelegant that we have these if else statements in three different places.
Can we share the code somehow?

Constellation replied:
Added `ScriptFetchParameters::parseType` and use it in these places.

rniwa commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -916,6 +917,24 @@ JSInternalPromise* GlobalObject::moduleLoaderImportModule(JSGlobalObject* global
+    if (!assertions.isEmpty()) {
+        ScriptFetchParameters::Type type = ScriptFetchParameters::Type::None;
+        auto typeString = assertions.get(vm.propertyNames->type.impl());
+        if (typeString == "json"_s)
```
Can we have a helper function which converts a type string to `ScriptFetchParameters::Type`?

Constellation replied:
Yeah, added `ScriptFetchParameters::parseType` :)

rniwa commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -1255,24 +1265,33 @@ JSInternalPromise* GlobalObject::moduleLoaderFetch(JSGlobalObject* globalObject,
-            promise->resolve(globalObject, sourceCode);
-            return promise;
-        }
+    if ((buffer.size() >= 4 && buffer[0] == '\0' && buffer[1] == 'a' && buffer[2] == 's' && buffer[3] == 'm') || (assertions && assertions->type() == ScriptFetchParameters::Type::WebAssembly)) {
```
Somewhat tangential but should this magic header check only be preformed when we're loading from the filesystem?
Otherwise, wouldn't we inadvertently execute random asm file serviced under another MIME?

Constellation replied:
Yes, but I think this is OK since it is `jsc` shell command, not related to WebContent process. And basically designed to be a test harness for JSC tests.

rniwa commented on Source/JavaScriptCore/parser/NodesAnalyzeModule.cpp:
```diff
@@ -32,6 +32,23 @@
+    // Currently, only "type" is supported.
+    std::optional<ScriptFetchParameters::Type> type;
+    for (auto& [key, value] : assertionList->assertions()) {
+        if (*key == vm.propertyNames->type)
+            type = ScriptFetchParameters::parseType(value->impl());
```
I guess we are to ignore unsupported keys here.

Constellation replied:
Yes, right now, we are hard-coding it because it is still not sure whether we will have platform-specific supported keys (so, not sure whether we need extensibility here). For now, keeping it, and later, if we start having yet another assertions, then, let's revisit. (I wonder if we end up always having the same supported assertions for every platforms JSC support, and probably it is fine).

rniwa commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -3602,11 +3602,13 @@ template <class TreeBuilder> typename TreeBuilder::ImportSpecifier Parser<LexerT
 template <typename LexerType>
 template <class TreeBuilder> typename TreeBuilder::ImportAssertionList Parser<LexerType>::parseImportAssertions(TreeBuilder& context)
 {
+    HashSet<UniquedStringImpl*> keys;
```
Can't we just use a boolean here?

Constellation replied:
We cannot do that since we need to detect duplicate keys.

rniwa commented on Source/JavaScriptCore/runtime/Completion.cpp:
```diff
@@ -266,4 +266,56 @@ JSInternalPromise* importModule(JSGlobalObject* globalObject, const Identifier&
+        String valueString = value.toWTFString(globalObject);
+        RETURN_IF_EXCEPTION(scope, { });
+
+        if (supportedAssertions.contains(key.impl()))
```
Using HashSet for this especially when there is exactly one item seems wasteful to me.
Why don't we just have a Vector instead?
It's probably a lot quicker to go through Vector of 3-4 items than going through a hashing.

Constellation replied:
OK, changed it to Vector.

rniwa reviewed (approved):
WebCore parts good to me but perhaps you want another JSC person review the JSC parts.

shvaikalesh reviewed (approved):
`Source/JavaScriptCore/*` part looks good to me.

Constellation committed:
[JSC] Implement import-assertion and JSON module
https://bugs.webkit.org/show_bug.cgi?id=243622

Reviewed by Alexey Shvayka and Ryosuke Niwa.

This patch implements import-assertions and JSON module.

1. Import assertions' syntax work was already done. We propagate assertions through ScriptFetchParameters, which already exists.
2. Get "type" assertion, and based on that, we parse JSON if it is "json".
3. Add SyntheticModuleRecord (this is described in the spec), and materialize JSON module as SyntheticModuleRecord.
4. Fix MIMETypeRegistry::isSupportedJSONMIMEType. "text/json" is explicitly listed in [1], so we should say true for that.
5. Dynamic import also gets assertions via its second parameter.

One test262 test is failing just because our harness is not handling it well (so this is not a bug in JSON module implementation).
WPT failures are due to old module loader's behavior (so, fixing it requires update of module loader implementation, not related to
JSON module itself).

[1]: https://mimesniff.spec.whatwg.org/#json-mime-type

* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/css-module/charset-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/css-module/content-type-checking-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/css-module/cors-crossorigin-requests-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/css-module/css-module-worker-test-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/css-module/import-css-module-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/css-module/integrity-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/css-module/load-error-events-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/import-assertions/empty-assertion-clause-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/import-assertions/invalid-type-assertion-error-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/import-assertions/unsupported-assertion-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/charset-bom.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/charset-bom.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/charset-bom.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/charset-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/cors-crossorigin-requests-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/integrity-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/json-module-service-worker-test.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/load-error-events-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/module-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/non-object.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/non-object.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/non-object.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/repeated-imports.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/repeated-imports.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/repeated-imports.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/scripting-1/the-script-element/json-module/valid-content-type-expected.txt:
* LayoutTests/platform/gtk/TestExpectations:
* Source/JavaScriptCore/API/JSAPIGlobalObject.mm:
(JSC::JSAPIGlobalObject::moduleLoaderImportModule):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/builtins/ModuleLoader.js:
(requestSatisfy):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::Heap):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/HeapSubspaceTypes.h:
* Source/JavaScriptCore/jsc.cpp:
(GlobalObject::moduleLoaderImportModule):
(GlobalObject::moduleLoaderFetch):
* Source/JavaScriptCore/parser/ModuleAnalyzer.cpp:
(JSC::ModuleAnalyzer::appendRequestedModule):
* Source/JavaScriptCore/parser/ModuleAnalyzer.h:
* Source/JavaScriptCore/parser/NodesAnalyzeModule.cpp:
(JSC::tryCreateAssertion):
(JSC::ImportDeclarationNode::analyzeModule):
(JSC::ExportAllDeclarationNode::analyzeModule):
(JSC::ExportNamedDeclarationNode::analyzeModule):
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser<LexerType>::parseImportAssertions):
* Source/JavaScriptCore/parser/SourceProvider.h:
* Source/JavaScriptCore/runtime/AbstractModuleRecord.cpp:
(JSC::AbstractModuleRecord::appendRequestedModule):
(JSC::AbstractModuleRecord::link):
(JSC::AbstractModuleRecord::evaluate):
(JSC::AbstractModuleRecord::dump):
* Source/JavaScriptCore/runtime/AbstractModuleRecord.h:
(JSC::AbstractModuleRecord::requestedModules const):
* Source/JavaScriptCore/runtime/Completion.cpp:
(JSC::retrieveAssertionsFromDynamicImportOptions):
* Source/JavaScriptCore/runtime/Completion.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::syntheticModuleRecordStructure const):
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSModuleLoader.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::JSONParse):
(JSC::JSONParseWithException):
* Source/JavaScriptCore/runtime/JSONObject.h:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/ScriptFetchParameters.h:
(JSC::ScriptFetchParameters::ScriptFetchParameters):
(JSC::ScriptFetchParameters::type const):
(JSC::ScriptFetchParameters::integrity const):
(JSC::ScriptFetchParameters::isTopLevelModule const):
(JSC::ScriptFetchParameters::create):
(JSC::ScriptFetchParameters::~ScriptFetchParameters): Deleted.
* Source/JavaScriptCore/runtime/SyntheticModuleRecord.cpp: Added.
(JSC::SyntheticModuleRecord::createStructure):
(JSC::SyntheticModuleRecord::create):
(JSC::SyntheticModuleRecord::SyntheticModuleRecord):
(JSC::SyntheticModuleRecord::destroy):
(JSC::SyntheticModuleRecord::finishCreation):
(JSC::SyntheticModuleRecord::visitChildrenImpl):
(JSC::SyntheticModuleRecord::link):
(JSC::SyntheticModuleRecord::evaluate):
(JSC::SyntheticModuleRecord::tryCreateWithExportNamesAndValues):
(JSC::SyntheticModuleRecord::tryCreateDefaultExportSyntheticModule):
(JSC::SyntheticModuleRecord::parseJSONModule):
* Source/JavaScriptCore/runtime/SyntheticModuleRecord.h: Added.
* Source/JavaScriptCore/wasm/js/JSWebAssemblyInstance.cpp:
(JSC::JSWebAssemblyInstance::tryCreate):
* Source/WebCore/bindings/js/CachedModuleScriptLoader.cpp:
(WebCore::CachedModuleScriptLoader::create):
(WebCore::CachedModuleScriptLoader::CachedModuleScriptLoader):
* Source/WebCore/bindings/js/CachedModuleScriptLoader.h:
* Source/WebCore/bindings/js/ModuleScriptLoader.h:
(WebCore::ModuleScriptLoader::parameters):
(WebCore::ModuleScriptLoader::ModuleScriptLoader):
* Source/WebCore/bindings/js/ScriptController.cpp:
(WebCore::ScriptController::loadModuleScriptInWorld):
(WebCore::ScriptController::loadModuleScript):
(WebCore::ScriptController::evaluateModule):
* Source/WebCore/bindings/js/ScriptController.h:
* Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
(WebCore::ScriptModuleLoader::fetch):
(WebCore::ScriptModuleLoader::importModule):
(WebCore::ScriptModuleLoader::notifyFinished):
* Source/WebCore/bindings/js/ScriptModuleLoader.h:
* Source/WebCore/bindings/js/WorkerModuleScriptLoader.cpp:
(WebCore::WorkerModuleScriptLoader::create):
(WebCore::WorkerModuleScriptLoader::WorkerModuleScriptLoader):
* Source/WebCore/bindings/js/WorkerModuleScriptLoader.h:
* Source/WebCore/bindings/js/WorkerScriptFetcher.h:
* Source/WebCore/dom/LoadableModuleScript.cpp:
(WebCore::LoadableModuleScript::LoadableModuleScript):
* Source/WebCore/dom/ModuleFetchParameters.h:
(WebCore::ModuleFetchParameters::create):
(WebCore::ModuleFetchParameters::ModuleFetchParameters):
(WebCore::ModuleFetchParameters::integrity const): Deleted.
(WebCore::ModuleFetchParameters::isTopLevelModule const): Deleted.
* Source/WebCore/platform/MIMETypeRegistry.cpp:
(WebCore::MIMETypeRegistry::isSupportedJSONMIMEType):
* Source/WebCore/workers/WorkerOrWorkletScriptController.cpp:
(WebCore::WorkerOrWorkletScriptController::loadAndEvaluateModule):
* Source/WebCore/workers/WorkerThread.cpp:
(WebCore::WorkerThread::evaluateScriptIfNecessary):
* Tools/TestWebKitAPI/Tests/WebCore/MIMETypeRegistry.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/253234@main

nicolo-ribaudo commented on Source/WebCore/bindings/js/ScriptModuleLoader.cpp:
```diff
@@ -299,15 +299,31 @@ static bool isWorkletOrServiceWorker(ScriptExecutionContext& context)
+        auto assertions = JSC::retrieveAssertionsFromDynamicImportOptions(&globalObject, parametersValue, { vm.propertyNames->type.impl() });
+        RETURN_IF_EXCEPTION(scope, { });
+        if (!assertions.isEmpty())
+            return JSC::ScriptFetchParameters::parseType(assertions.get(vm.propertyNames->type.impl())).value_or(JSC::ScriptFetchParameters::Type::JavaScript);
```
If `JSC::ScriptFetchParameters::parseType` returns `nullopt`, this shouldn't fallback to `JSC::ScriptFetchParameters::Type::JavaScript` but reject the promise.

Step 5 of https://html.spec.whatwg.org/#fetch-an-import()-module-script-graph says
> If the result of running the [module type allowed](https://html.spec.whatwg.org/#module-type-allowed) steps given _moduleType_ and _settings object_ is false, then run _onComplete_ given null, and return.

(until a few days ago, "run onComplete given null" was written as "asynchronously complete this algorithm with null".

And then step 6.2 of https://html.spec.whatwg.org/#hostimportmoduledynamically(referencingscriptormodule,-modulerequest,-promisecapability) rejects the promise if the result is null:
> If result is null, then set _promise_ to [a promise rejected with](https://webidl.spec.whatwg.org/#a-promise-rejected-with) a new [TypeError](https://tc39.es/ecma262/#sec-native-error-types-used-in-this-standard-typeerror).

I found this in https://github.com/web-platform-tests/wpt/pull/35949
